### PR TITLE
#13 read file from bucketfs inside a udf

### LINKED ISSUE #13: Read file from bucketfs inside a UDF from `/buckets`
## Background

- The BucketFS is in Exasol mounted as read only filesystem on each node
- It is faster to read files in the BucketFS from the filesystem as downloading them via HTTP

## Acceptance Criteria

- Read a file from the BucketFS via the file system given a Bucket Object and a file path
- Targets are String, FileObj, to file, to object via joblib

### PR DESCRIPTION
**Pull request by MarleneKress79789**
Closes #13

### CODE REVIEW

File: exasol_bucketfs_utils_python/bucketfs_location.py [hunk 1]
```diff
@@ -13,6 +14,8 @@ class BucketFSLocation(AbstractBucketFSLocation):
     BucketFSLocation implements AbstractBucketFSLocation.
     BucketFSLocation is used to upload fileobjects, strings or joblib objects to the BucketFS given a path and the object,
     or to download objects into strings, fileobjects or joblib objects from the BucketFS given a file path.
+    Also able to read files from the BucketFS directly, if called from inside a UDF.
```
Comment (tomuben) [inline exasol_bucketfs_utils_python/bucketfs_location.py:17]:
```suggestion
    Also able to read files from the BucketFS directly, if called from inside an UDF.
```

File: exasol_bucketfs_utils_python/bucketfs_location.py [hunk 2]
```diff
@@ -13,6 +14,8 @@ class BucketFSLocation(AbstractBucketFSLocation):
...
+    If reading an object via joblib inside an UDF, make sure the object type is known inside the UDF.
```
Comment (tomuben) [inline exasol_bucketfs_utils_python/bucketfs_location.py:18]:
```suggestion
    If reading an object via joblib inside of an UDF, make sure the object type is known inside the UDF.
```

File: exasol_bucketfs_utils_python/bucketfs_location.py [hunk 3]
```diff
@@ -59,3 +62,34 @@ def upload_fileobj_to_bucketfs(self,
...
+
+    def read_file_from_bucketfs_via_joblib(self, bucket_file_path: str) -> typing.Any:
+        result = from_BFS.read_file_from_bucketfs_via_joblib(
+            self.get_complete_file_path_in_bucket(bucket_file_path),
+            self.bucket_config
+        )
+        return result
+
+
+    def read_file_from_bucketfs_to_file(self, bucket_file_path: str, local_file_path: Path):
```
Comment (tomuben) [inline exasol_bucketfs_utils_python/bucketfs_location.py:81]: Add type hint of return value

File: exasol_bucketfs_utils_python/bucketfs_location.py [hunk 4]
```diff
@@ -59,3 +62,34 @@ def upload_fileobj_to_bucketfs(self,
...
+        result = from_BFS.read_file_from_bucketfs_to_file(
+            self.get_complete_file_path_in_bucket(bucket_file_path),
+            self.bucket_config,
+            local_file_path
+        )
+        return result
+
+    def read_file_from_bucketfs_to_fileobj(self, bucket_file_path: str, fileobj: typing.IO):
```
Comment (tomuben) [inline exasol_bucketfs_utils_python/bucketfs_location.py:89]: Add type hint of return value

File: exasol_bucketfs_utils_python/load_file_from_local_fs.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+import typing
+from pathlib import Path
+from tempfile import NamedTemporaryFile
+import joblib
+
+
+def read_file_from_bucketfs_to_string(bucket_file_path: str, bucket_config: BucketConfig) -> str:
+    """
+    Read a file from the specified path in the bucket in the BucketFs into a string.
+    Can be used inside a UDF.
```
Comment (tomuben) [inline exasol_bucketfs_utils_python/load_file_from_local_fs.py:12]:
```suggestion
    Can be used inside of an UDF.
```

File: exasol_bucketfs_utils_python/load_file_from_local_fs.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+    :param bucket_config: BucketConfig for the bucket to read from
+    :param bucket_file_path: Path in the bucket to the file to read, given as a string.
+    :return: The content of the file in the BucketFS as string
+    """
+    if bucket_file_path is None:
+        raise ValueError("bucket_file_path can't be None")
+    bucket_path = generate_bucket_udf_path(bucket_config, bucket_file_path)
+    text_as_string = open(bucket_path).read()
```
Comment (tomuben) [inline exasol_bucketfs_utils_python/load_file_from_local_fs.py:21]: Maybe better to use "with open(bucket_path) as bucker_file:..."

File: exasol_bucketfs_utils_python/load_file_from_local_fs.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    return text_as_string
+
+
+def read_file_from_bucketfs_to_file(bucket_file_path: str, bucket_config: BucketConfig, local_file_path: Path):
```
Comment (tomuben) [inline exasol_bucketfs_utils_python/load_file_from_local_fs.py:25]: Add type hint of return value

File: exasol_bucketfs_utils_python/load_file_from_local_fs.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    :param bucket_config: BucketConfig for the bucket to download from
+    :param local_file_path: Path in the bucket to save the file content in.
+    :param bucket_file_path: Path in the bucket to the file to read, given as a string.
+    :return: None
+    """
+    with local_file_path.open("wb") as f:
+        read_file_from_bucketfs_to_fileobj(bucket_file_path, bucket_config, fileobj=f)
+
+
+def read_file_from_bucketfs_to_fileobj(bucket_file_path: str, bucket_config: BucketConfig, fileobj: typing.IO):
```
Comment (tomuben) [inline exasol_bucketfs_utils_python/load_file_from_local_fs.py:39]: type hint of return value

File: exasol_bucketfs_utils_python/load_file_from_local_fs.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        raise ValueError("bucket_file_path can't be None")
+    bucket_path = generate_bucket_udf_path(bucket_config, bucket_file_path)
+    text_as_string = open(bucket_path).read()
+    return text_as_string
+
+
+def read_file_from_bucketfs_to_file(bucket_file_path: str, bucket_config: BucketConfig, local_file_path: Path):
+    """
+    Read a file from the specified path in the bucket in the BucketFs and save as a local file
+    Can be used inside a UDF.
```
Comment (tomuben) [inline exasol_bucketfs_utils_python/load_file_from_local_fs.py:28]:
```suggestion
    Can be used inside of an UDF.
```

File: tests/fixtures/build_language_container_fixture.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    script_path = None
+    while current_path != current_path.root:
+        script_path = Path(current_path, script_name)
+        if script_path.exists():
+            break
+        current_path = current_path.parent
+    if script_path.exists():
+        return script_path
+    else:
+        raise RuntimeError("Could not find build_language_container.sh")
```
Comment (tomuben) [inline tests/fixtures/build_language_container_fixture.py:18]:
```suggestion
        raise RuntimeError(f"Could not find {script_name}")
```

File: tests/fixtures/upload_language_container_fixture.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    container_bucketfs_location = \
+        bucket_fs_factory.create_bucketfs_location(
+            url=container_connection.address,
+            user=container_connection.user,
+            pwd=container_connection.password,
+            base_path=None)
+    container_path = Path(language_container["container_path"])
+    alter_session = Path(language_container["alter_session"])
+    pyexasol_connection.execute(f"ALTER SYSTEM SET SCRIPT_LANGUAGES='{alter_session}'")
+    pyexasol_connection.execute(f"ALTER SESSION SET SCRIPT_LANGUAGES='{alter_session}'")
```
Comment (tomuben) [inline tests/fixtures/upload_language_container_fixture.py:36]: Any particular reason you call both: ALTER SYSYEM AND ALTER SESSION?
If so, please comment why.

File: tests/test_bucketfs_location.py [hunk 12]
```diff
@@ -29,15 +35,255 @@ def __eq__(self, other):
...
     bucketfs_location_upload.upload_object_to_bucketfs_via_joblib(test_value, bucket_file_path)
     result = bucketfs_location_download.download_object_from_bucketfs_via_joblib(bucket_file_path)
     assert result == test_value
+    delete_testfile_from_BucketFS(file_path=str(bucket_base_path) + "/" + bucket_file_path,
+                                  bucket_config=bucketfs_location_upload.bucket_config)
+
+
+@pytest.mark.usefixtures("upload_language_container",
+                             "pyexasol_connection",
+                             "db_connection")
```
Comment (tomuben) [inline tests/test_bucketfs_location.py:56]: You use the fixture db_connection, but you only use pyexasol_connection

File: tests/test_bucketfs_location.py [hunk 13]
```diff
@@ -29,15 +35,255 @@ def __eq__(self, other):
...
+            f"""select {target_schema}."LoadFromLocalFS"('{bucketfs_location_read.bucket_config}','{bucket_file_path}'
+            )""").fetchall()
+        assert output_test_string[0][0] == test_string
+    finally:
+        delete_testfile_from_BucketFS(file_path=str(bucket_base_path) + "/" + bucket_file_path,
+                                      bucket_config=bucketfs_location_read.bucket_config)
+        pyexasol_connection.execute(f"DROP SCHEMA IF EXISTS {target_schema} CASCADE;")
+
+
+def test_read_files_via_joblib_from_bucketfs_inside_udf(upload_language_container, pyexasol_connection):
```
Comment (tomuben) [inline tests/test_bucketfs_location.py:114]: I am not sure, but don't you have to indicate usage of the fixtures here, too?
```
@pytest.mark.usefixtures("upload_language_container",
                             "pyexasol_connection")
```

File: tests/test_bucketfs_location.py [hunk 14]
```diff
@@ -29,15 +35,255 @@ def __eq__(self, other):
...
+            f"""select {target_schema}."LoadFromLocalFS"('{bucketfs_location_read.bucket_config}','{bucket_file_path}'
+            )""").fetchall()
+        assert output_test_string[0][0] == test_python_object.bucketfs_name
+    finally:
+        delete_testfile_from_BucketFS(file_path=str(bucket_base_path) + "/" + bucket_file_path,
+                                      bucket_config=bucketfs_location_read.bucket_config)
+        pyexasol_connection.execute(f"DROP SCHEMA IF EXISTS {target_schema} CASCADE;")
+
+
+def test_read_files_to_file_from_bucketfs_inside_udf(upload_language_container, pyexasol_connection):
```
Comment (tomuben) [inline tests/test_bucketfs_location.py:173]: I am not sure, but don't you have to indicate usage of the fixtures here, too?
```
@pytest.mark.usefixtures("upload_language_container",
                             "pyexasol_connection")
```

File: tests/test_bucketfs_location.py [hunk 15]
```diff
@@ -29,15 +35,255 @@ def __eq__(self, other):
...
+        output_test_string = pyexasol_connection.execute(
+            f"""select {target_schema}."LoadFromLocalFS"('{bucketfs_location_read.bucket_config}','{bucket_file_path}'
+            )""").fetchall()
+        assert output_test_string[0][0] == test_string
+    finally:
+        delete_testfile_from_BucketFS(file_path=str(bucket_base_path) + "/" + bucket_file_path,
+                                      bucket_config=bucketfs_location_read.bucket_config)
+        pyexasol_connection.execute(f"DROP SCHEMA IF EXISTS {target_schema} CASCADE;")
+
+def test_read_files_to_fileobj_from_bucketfs_inside_udf(upload_language_container, pyexasol_connection):
```
Comment (tomuben) [inline tests/test_bucketfs_location.py:230]: I am not sure, but don't you have to indicate usage of the fixtures here, too?
```
@pytest.mark.usefixtures("upload_language_container",
                             "pyexasol_connection")
```

**Review by tomuben — COMMENTED** [review]
Please check my comments.
Also, what's the idea behind the mock? Is it used in integration tests in other projects?

**Comment by MarleneKress79789** [comment]
> Also, what's the idea behind the mock? Is it used in integration tests in other projects?

Torsten uses the Mock for fast testing when he does not want to set up a database in the background.

**Comment by tkilias** [comment]
> > Also, what's the idea behind the mock? Is it used in integration tests in other projects?
> 
> Torsten uses the Mock for fast testing when he does not want to set up a database in the background.

Yep, I use it together with the udf-mock-python. We need the bucketfs location mock here, because it is instantiated via the Bucketfs factory, which works as injection point of the Mock for UDFs, because there is not other way to inject a object into UDFs.

File: exasol_bucketfs_utils_python/bucketfs_location.py (hunk 3 above)
Comment (MarleneKress79789) [reply exasol_bucketfs_utils_python/bucketfs_location.py]: done

File: exasol_bucketfs_utils_python/bucketfs_location.py (hunk 4 above)
Comment (MarleneKress79789) [reply exasol_bucketfs_utils_python/bucketfs_location.py]: done

File: exasol_bucketfs_utils_python/load_file_from_local_fs.py (hunk 5 above)
Comment (MarleneKress79789) [reply exasol_bucketfs_utils_python/load_file_from_local_fs.py]: done

File: exasol_bucketfs_utils_python/load_file_from_local_fs.py (hunk 6 above)
Comment (MarleneKress79789) [reply exasol_bucketfs_utils_python/load_file_from_local_fs.py]: done

File: exasol_bucketfs_utils_python/load_file_from_local_fs.py (hunk 7 above)
Comment (MarleneKress79789) [reply exasol_bucketfs_utils_python/load_file_from_local_fs.py]: done

File: tests/test_bucketfs_location.py (hunk 12 above)
Comment (MarleneKress79789) [reply tests/test_bucketfs_location.py]: done

File: tests/fixtures/upload_language_container_fixture.py (hunk 11 above)
Comment (MarleneKress79789) [reply tests/fixtures/upload_language_container_fixture.py]: No, i just took that from the data science project. Is changed now.

File: tests/fixtures/build_language_container_fixture.py (hunk 10 above)
Comment (MarleneKress79789) [reply tests/fixtures/build_language_container_fixture.py]: done

File: exasol_bucketfs_utils_python/load_file_from_local_fs.py (hunk 9 above)
Comment (MarleneKress79789) [reply exasol_bucketfs_utils_python/load_file_from_local_fs.py]: done

File: exasol_bucketfs_utils_python/bucketfs_location.py (hunk 2 above)
Comment (MarleneKress79789) [reply exasol_bucketfs_utils_python/bucketfs_location.py]: done

File: build_language_container.sh [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+echo "Copy" dist/*.whl "$RELEASE_BUILD_STEP_DIST_DIRECTORY"
+mkdir "$RELEASE_BUILD_STEP_DIST_DIRECTORY" || true
+cp dist/*.whl "$RELEASE_BUILD_STEP_DIST_DIRECTORY"
+trap 'rm -rf "$RELEASE_BUILD_STEP_DIST_DIRECTORY"' EXIT
+
+echo "Build container"
+./language_container/exaslct export --flavor-path language_container/exasol-bucketfs-utils-python-container/
+
+echo "Generate language activation"
+./language_container/exaslct generate-language-activation --flavor-path language_container/exasol-bucketfs-utils-python-container/ --bucketfs-name bfsdefault --bucket-name default --path-in-bucket container --container-name ml
```
Comment (tomuben) [inline build_language_container.sh:19]: This file is only used for testing, maybe we should move it to the respective folder.

**Review by tomuben — COMMENTED** [review]
Please check my comment

File: build_language_container.sh (hunk 16 above)
Comment (tkilias) [reply build_language_container.sh:19]: If you move it, please be aware that you might need to change the fixture which looks for it

Comment (MarleneKress79789) [reply build_language_container.sh:19]: If I move the file, the .build_output directory is generated inside the test directory. I'm having trouble finding where to change this.

Comment (MarleneKress79789) [reply build_language_container.sh:19]: It works like this, but i don't think this is what we want.

Comment (MarleneKress79789) [reply build_language_container.sh]: fixed

File: tests/build_language_container.sh [hunk 17]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+SCRIPT_DIR="$(dirname "$(readlink -f "${BASH_SOURCE[0]}")")"
+pushd  $SCRIPT_DIR &> /dev/null
+poetry build
+
+
+RELEASE_BUILD_STEP_DIST_DIRECTORY="./../language_container/exasol-bucketfs-utils-python-container/flavor_base/release/dist"
```
Comment (tomuben) [inline tests/build_language_container.sh:10]:
```suggestion
RELEASE_BUILD_STEP_DIST_DIRECTORY="$SCRIPT_DIR/../language_container/exasol-bucketfs-utils-python-container/flavor_base/release/dist"
```

Thus, it's independent of the from where you call the script

File: tests/build_language_container.sh [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+echo "Copy" ./../dist/*.whl "$RELEASE_BUILD_STEP_DIST_DIRECTORY"
```
Comment (tomuben) [inline tests/build_language_container.sh:11]: Same as above

File: tests/build_language_container.sh [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+mkdir "$RELEASE_BUILD_STEP_DIST_DIRECTORY" || true
+cp ./../dist/*.whl "$RELEASE_BUILD_STEP_DIST_DIRECTORY"
```
Comment (tomuben) [inline tests/build_language_container.sh:13]: Same as above

File: tests/build_language_container.sh [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+trap 'rm -rf "$RELEASE_BUILD_STEP_DIST_DIRECTORY"' EXIT
+
+echo "Build container"
+./../language_container/exaslct export --flavor-path ./../language_container/exasol-bucketfs-utils-python-container/  --output-directory ../.build_output
+
+echo "Generate language activation"
+./../language_container/exaslct generate-language-activation --flavor-path ./../language_container/exasol-bucketfs-utils-python-container/ --bucketfs-name bfsdefault --bucket-name default --path-in-bucket container --container-name ml
```
Comment (tomuben) [inline tests/build_language_container.sh:20]: Same as above

**Review by tomuben — APPROVED** [review]

**Pull request merged**